contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,325 @@
+        game.playerTwo = msg.sender;
+        game.playerTwoMoveHash = moveHash;
+        game.step = uint256(Steps.PLAYER_TWO_MOVE);
+        game.expiryDate = expiryDate;
+        depositStake(game.stake);
+
+        emit PlayerTwoMoves(msg.sender, msg.value, gameKey, moveHash, expiryDate);
+    }
+
+    function depositStake(uint256 stake) internal {
+
+        if (msg.value == stake) {
+            return;
+        }
+        uint256 senderBalance = balances[msg.sender];
+
+        if (msg.value < stake) {
+            require(senderBalance >= stake.sub(msg.value), "Insufficient balance");
+            balances[msg.sender] = senderBalance.sub(stake.sub(msg.value));
+        } else {
+            balances[msg.sender] = senderBalance.add(msg.value).sub(stake);
+        }
+    }
+
+    function revealPlayerOne(bytes32 secret, uint256 playerOneMove) external whenNotPaused {
+
+        bytes32 gameKey = createPlayerOneMoveHash(secret, playerOneMove);
+        Game storage game = games[gameKey];
+        require(game.step == uint256(Steps.PLAYER_TWO_MOVE), INVALID_STEP_MSG);
+
+        uint256 expiryDate = block.timestamp.add(FORFEIT_WINDOW);
+        game.playerOneMove = playerOneMove;
+        game.step = uint256(Steps.PLAYER_ONE_REVEAL);
+        game.expiryDate = expiryDate;
+
+        emit PlayerOneReveals(msg.sender, gameKey, playerOneMove, expiryDate);
+    }
+
+    function revealPlayerTwo(bytes32 gameKey, bytes32 secret, uint256 playerTwoMove) external whenNotPaused {
+
+        Game storage game = games[gameKey];
+        require(game.step == uint256(Steps.PLAYER_ONE_REVEAL), INVALID_STEP_MSG);
+        address playerTwo = game.playerTwo;
+
+        //Validate move
+        bytes32 expectedMoveHash = createPlayerTwoMoveHash(gameKey, secret, playerTwoMove);
+        require(game.playerTwoMoveHash == expectedMoveHash, HASH_MISMATCH_MSG);
+
+        emit PlayerTwoReveals(msg.sender, gameKey, playerTwoMove);
+
+        uint256 stake = game.stake;
+        address playerOne = game.playerOne;
+        uint256 playerOneMove = game.playerOneMove;
+
+        //Draw
+        if (playerOneMove == playerTwoMove){
+            resetGame(game);
+            balances[playerOne] = balances[playerOne].add(stake);
+            balances[playerTwo] = balances[playerTwo].add(stake);
+            DrawRefund(playerOne, playerTwo, gameKey, stake, stake);
+
+            return;
+        }
+
+        //Determine and reward winner
+        address winner;
+        address loser;
+
+        if (isWinner(playerOneMove, playerTwoMove)) {
+            winner = playerOne;
+            loser = playerTwo;
+        } else {
+            winner = playerTwo;
+            loser = playerOne;
+        }
+        uint256 fee = stake.mul(FEE_PERCENTAGE).div(100);
+        uint256 winnings = stake.mul(2).sub(fee);
+        balances[winner] = balances[winner].add(winnings);
+        emit GamePayment(winner, gameKey, winnings);
+
+        balances[loser] = balances[loser].add(fee);
```

Comment:
I don't understand the logic to have a fee for the loser.

Reply:
I wanted to provide a small payment incentive for player two to reveal even if they know before that they have lost because player one has already revealed.

It only seemed fair to provide that incentive to any losing player.

Perhaps over-engineering?

Reply:
Thinking about it, the other player perhaps would prefer simply to collect the full amount in the forfeit without the fees. However, in duration of the time window the ETH value could fluctuate so potentially more risk than getting slightly less winnings much earlier?

Reply:
I think it is too much engineering too early.